### PR TITLE
Reserve variable starting characters.

### PR DESCRIPTION
**Pull request by Moderocky**
### Description

In an upcoming version, we're looking at instituting new special variable types (e.g. unsaved, reference).
No plans are finalised for this, but the first step is reserving the starting characters (like `_` for local variables).

The current characters reserved are:
```applescript
{~variable}
{.variable}
{$variable}
{!variable}
{&variable}
{^variable}
{*variable} # See note 2
{+variable}
{-variable} # See note 1
```

Currently, this will only issue a _warning_ when they're used (but it won't prevent anything loading).
Presumably, in some future version, this will become an error (or the variable will have some special functionality)!

### Why these characters?

It was a pretty arbitrary choice, we just picked a handful of special characters to give us some freedom for future changes.

## Note 1: `-` minus variable

A lot of people use the `(?!-).*` pattern in the config to prevent `{-variable}` from being saved **at all** (i.e. to vanish when Skript terminates).

We'd like to institute proper behaviour for an 'ephemeral variable' in the future (rather than just using the config pattern as a patchwork), so that's why we're reserving this character.

However, I've tried to skip the warning **IFF** you're using this pattern in the config (since there's nothing to warn you about).

## Note 2: `*` star

Stars are already prohibited in variable names (other than in a `list::*`) so this one shouldn't affect anybody.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

**Review by sovdeeth — APPROVED** [review]
maybe we could add a suppression option but i don't mind either way

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/variables/Variables.java [hunk 1]
```diff
@@ -144,6 +144,13 @@ public Class<? extends ConfigurationSerializable> getClass(@NonNull String id) {
 	 */
 	static final List<VariablesStorage> STORAGES = new ArrayList<>();
 
+	/**
+	 * @return a copy of the list of variable storage handlers
+	 */
+	public static List<VariablesStorage> getStores() {
+		return Collections.unmodifiableList(STORAGES);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/variables/Variables.java:151]: Might be worth annotating this as Unmodifiable or Unmodifiable View. I don't recall whether this method creates a copy or a view.

File: src/main/java/ch/njol/skript/lang/Variable.java [hunk 2]
```diff
@@ -119,7 +125,26 @@ private Variable(VariableString name, Class<? extends T>[] types, boolean local,
 	 * @return true if the name is valid, false otherwise.
 	 */
 	public static boolean isValidVariableName(String name, boolean allowListVariable, boolean printErrors) {
-		name = name.startsWith(LOCAL_VARIABLE_TOKEN) ? "" + name.substring(LOCAL_VARIABLE_TOKEN.length()).trim() : "" + name.trim();
+		assert !name.isEmpty(): "Variable name should not be empty";
+		char first = name.charAt(0);
+		check_reserved_tokens:
+		for (char token : reservedTokens) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Variable.java:131]:
```suggestion
		check_reserved_tokens: for (char token : reservedTokens) {
```
I feel like it's clearer to typically combine these, but I'm not sure what the code conventions suggest (if anything)

File: src/main/java/ch/njol/skript/lang/Variable.java [hunk 3]
```diff
@@ -119,7 +125,26 @@ private Variable(VariableString name, Class<? extends T>[] types, boolean local,
...
+				*/
+				if (first == '-') {
+					for (VariablesStorage store : Variables.getStores()) {
+						@Nullable Pattern pattern = store.getNamePattern();
+						if (pattern != null && pattern.pattern().equals("(?!-).*"))
+							continue check_reserved_tokens;
+					}
+				}
+				Skript.warning("The character '" + token + "' is reserved at the start of variable names, " +
+								   "and may be restricted in future versions");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Variable.java:144]: It appears that tabs and spaces are being mixed on this line. Two standard indents should suffice

File: src/main/java/ch/njol/skript/lang/Variable.java [hunk 4]
```diff
@@ -55,21 +70,12 @@
...
-import java.util.Map.Entry;
-import java.util.NoSuchElementException;
-import java.util.TreeMap;
-
 public class Variable<T> implements Expression<T> {
 
 	private final static String SINGLE_SEPARATOR_CHAR = ":";
 	public final static String SEPARATOR = SINGLE_SEPARATOR_CHAR + SINGLE_SEPARATOR_CHAR;
 	public final static String LOCAL_VARIABLE_TOKEN = "_";
+	private static final char[] reservedTokens = {'~', '.', '+', '$', '!', '&', '^', '*', '-'};
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Variable.java:78]: Is there any concern about these contents being modified (even via reflection potentially)

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
These are generally my own thoughts, so please feel free to counter back

File: src/main/java/ch/njol/skript/lang/Variable.java (hunk 2 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/lang/Variable.java:131]: Reformat puts it on a new line so I usually put it on a new line, up to you though.

Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/lang/Variable.java:131]: It's fine as is if it's just the reformatter

File: src/main/java/ch/njol/skript/lang/Variable.java (hunk 4 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/lang/Variable.java:78]: I don't think this is a particularly high-security risk 😔

File: src/main/java/ch/njol/skript/lang/Variable.java [hunk 5]
```diff
@@ -119,7 +125,26 @@ private Variable(VariableString name, Class<? extends T>[] types, boolean local,
...
+				A lot of people already use '-' so we want to skip this warning iff they're using it here
+				*/
+				if (first == '-') {
+					for (VariablesStorage store : Variables.getStores()) {
+						@Nullable Pattern pattern = store.getNamePattern();
+						if (pattern != null && pattern.pattern().equals("(?!-).*"))
+							continue check_reserved_tokens;
+					}
+				}
+				Skript.warning("The character '" + token + "' is reserved at the start of variable names, " +
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Variable.java:143]: I would like to see this message in `english.lang`. I think it is reasonable for us to provide translation support for these internal messages.

**Review by APickledWalrus — APPROVED** [review]
Looks good. I have a small request

**Pull request merged**